File: Solutions/Neetcode150/Stack/5-CarFleet/solution.ts
function carFleet(target: number, position: number[], speed: number[]): number {
  const positionToSpeed = new Map<number, number>(); // { position: speedAtThatPosition }
  for (let i = 0; i < position.length; i++) {
    positionToSpeed.set(position[i], speed[i]);
  }

  const positionDescendingOrder = position.sort((a, b) => b - a);
  const fleetStack: number[] = [];

  for (const position of positionDescendingOrder) {
    const timeToTarget = (target - position) / positionToSpeed.get(position)!;

    if (
      !fleetStack.length ||
      fleetStack[fleetStack.length - 1] < timeToTarget
    ) {
      fleetStack.push(timeToTarget);
    }
  }

  return fleetStack.length;
}

File: Solutions/Neetcode150/Stack/5-CarFleet/solution.py
from typing import List


class Solution:
    def carFleet(self, target: int, position: List[int], speed: List[int]) -> int:
        pos_to_speed = {}
        for i, p in enumerate(position):
            pos_to_speed[p] = speed[i]

        position.sort(reverse=True)

        stack = []
        for p in position:
            s = pos_to_speed[p]
            time_to_target = (target - p) / s

            if not stack or (stack and stack[-1] < time_to_target):
                stack.append(time_to_target)

        return len(stack)
